query/src/main/scala/filodb/query/exec/ExecPlan.scala

```diff
@@ -80,23 +80,167 @@ trait ExecPlan extends QueryCommand {
+            // transformers that cannot handle empty results
+            (acc._1, resultSchema)
+          } else {
+            val rangeVector : Observable[RangeVector] = transf.apply(
+              acc._1, querySession, queryContext.plannerParams.sampleLimit, acc._2, paramRangeVector
+            )
+            val schema = transf.schema(resultSchema)
+            (rangeVector, schema)
+          }
+        }
+        if (finalRes._2 == ResultSchema.empty) {
+          Observable.fromIterable(Seq(StrQueryResultHeader(queryContext.queryId, finalRes._2),
+            StrQueryResultFooter(queryContext.queryId, querySession.queryStats,
+              querySession.resultCouldBePartial, querySession.partialResultsReason)))
+        } else {
+          val recSchema = SerializedRangeVector.toSchema(finalRes._2.columns, finalRes._2.brSchemas)
+          Kamon
+            .histogram(
+              "query-execute-time-elapsed-step2-transformer-pipeline-setup",
+              MeasurementUnit.time.milliseconds
+            )
+            .withTag("plan", getClass.getSimpleName)
+            .record(Math.max(0, System.currentTimeMillis - startExecute))
+          makeResult(finalRes._1, recSchema, finalRes._2)
+        }
+      }
+      Observable.fromTask(task).flatten
+    }
+
+    def makeResult(rv: Observable[RangeVector], recordSchema: RecordSchema,
+                   resultSchema: ResultSchema): Observable[StrQueryResponse] = {
+      @volatile var numResultSamples = 0 // BEWARE - do not modify concurrently!!
+      @volatile var resultSize = 0L
+      val queryResults = rv.doOnStart(_ => Task.eval(span.mark("before-first-materialized-result-rv")))
+        .map {
+          case srvable: SerializableRangeVector => srvable
+          case rv: RangeVector =>
+            // materialize, and limit rows per RV
+            val execPlanString = queryWithPlanName(queryContext)
+            val builder = SerializedRangeVector.newBuilder()
+            val srv = SerializedRangeVector(rv, builder, recordSchema, execPlanString)
+            if (rv.outputRange.isEmpty)
+              qLogger.debug(s"Empty rangevector found. Rv class is:  ${rv.getClass.getSimpleName}, " +
+                s"execPlan is: $execPlanString, execPlan children ${this.children}")
+            srv
+        }
+        .map { srv =>
+          numResultSamples += srv.numRowsSerialized
+          // fail the query instead of limiting range vectors and returning incomplete/inaccurate results
+          if (enforceSampleLimit && numResultSamples > queryContext.plannerParams.sampleLimit)
+            throw new BadQueryException(s"This query results in more than ${queryContext.plannerParams.
+              sampleLimit} samples. Try applying more filters or reduce time range.")
+
+          resultSize += srv.estimateSerializedRowBytes + srv.key.keySize
+          if (resultSize > queryContext.plannerParams.resultByteLimit) {
```

Comment:
If I understand correctly: this limit isn't enforced while data is streamed? Can we enforce this size limit inside `ActorPlanDispatcher` during the stream (i.e. once N bytes have been streamed, we throw an exception)?

Reply:
Can we also stream data from C* and enforce a size limit on that stream, too?

Reply:
Yes - query limits are a TODO. I took a shortcut here for the POC.
Cassandra paging is already a stream today. See [code here](https://github.com/filodb/FiloDB/blob/735bc0faf5784d53609fce1251bafed18ad0c25d/core/src/main/scala/filodb.core/memstore/OnDemandPagingShard.scala#L147)